En este archivo puedes escribir lo que estimes conveniente. Te recomendamos detallar tu solución y todas las suposiciones que estás considerando. Aquí puedes ejecutar las funciones que definiste en los otros archivos de la carpeta src, medir el tiempo, memoria, etc.

## Preparacion Ambiente de trabajo.

1. Entorno de trabajo: La solucion del desafio es desarrollada en un workbench de Vertex AI, la instancia utilizada es una e2-standar-2 de 2 CPU virtuales y 8 GB de RAM con python3 y conda previamente instalado.

2. Para el manejo de versiones del desarrollo, se utilizara la herramienta GitHub, para la configuracion de este ambiente de trabajo se siguio los siguientes pasos:

    a. Se descargo y se descomprimio la carpeta challenge_DE, la cual contiene la estructura del proyecto, brindada por el equipo de LATAM.
    
    b. Inicializacion GIT en la carpeta del proyecto: 
    - *git init*
    
    
    c. Agregar los archivos al repositorio:
    - *git add .*
    
    d. Realizacion del primer commit para registrar los archivos iniciales de repo:
    - *git commit -m "Initial commit w/ project skeleton*
    
    
    e. Atenticacion con mi cuenta de GitHub personal:
    - *gh auth login*
    
    
    f. Creacion de un repositorio en GitHub, publico.
    - *gh repo create twitter_chall --public --confirm*
    
    
    g. Coneccion de mi repositorio local al repositorio remoto y realizacion del primer push de los cambios.
    - *git remote add origin https://github.com/dcharry/twitter_chall.git*
    - *git branch -M main*
    - *git push -u origin main*

In [2]:
import pandas as pd
import numpy as np

## Lectura data de twitter

1. En la terminal se installo la biblioteca gdwon para descargar el archivo desde google drive

2. Se creo la carpeta src/data para descargar la data a trabajar.

3. se corre el comando 'gdwon --id' para descargar el archivo usando el ID del drive, el cual es '1ig2ngoXFTxP5Pa8muXo02mDTFexZzsis' 


4. se descomprime el archivo utilizando el comando unzip en la terminal.

5. El contenido es un archivo json, se utiliza la libreria pandas para abrir el archivo en formato panda.DataFrame y explorar su contenido. (Primer acercamiento a la data)





In [3]:
df = pd.read_json('data/farmers-protest-tweets-2021-2-4.json', lines = True)


In [4]:
df.head()

url  \
0  https://twitter.com/ArjunSinghPanam/status/136...   
1  https://twitter.com/PrdeepNain/status/13645062...   
2  https://twitter.com/parmarmaninder/status/1364...   
3  https://twitter.com/anmoldhaliwal/status/13645...   
4  https://twitter.com/KotiaPreet/status/13645061...   

                       date  \
0 2021-02-24 09:23:35+00:00   
1 2021-02-24 09:23:32+00:00   
2 2021-02-24 09:23:22+00:00   
3 2021-02-24 09:23:16+00:00   
4 2021-02-24 09:23:10+00:00   

                                             content  \
0  The world progresses while the Indian police a...   
1  #FarmersProtest \n#ModiIgnoringFarmersDeaths \...   
2  ਪੈਟਰੋਲ ਦੀਆਂ ਕੀਮਤਾਂ ਨੂੰ ਮੱਦੇਨਜ਼ਰ ਰੱਖਦੇ ਹੋਏ \nਮੇ...   
3  @ReallySwara @rohini_sgh watch full video here...   
4  #KisanEktaMorcha #FarmersProtest #NoFarmersNoF...   

                                     renderedContent                   id  \
0  The world progresses while the Indian police a...  1364506249291784198   
1  #FarmersProtest \n#ModiIgnoringFarmersDeaths \...  1364506237451313155   
2  ਪੈਟਰੋਲ ਦੀਆਂ ਕੀਮਤਾਂ ਨੂੰ ਮੱਦੇਨਜ਼ਰ ਰੱਖਦੇ ਹੋਏ \nਮੇ...  1364506195453767680   
3  @ReallySwara @rohini_sgh watch full video here...  1364506167226032128   
4  #KisanEktaMorcha #FarmersProtest #NoFarmersNoF...  1364506144002088963   

                                                user  \
0  {'username': 'ArjunSinghPanam', 'displayname':...   
1  {'username': 'PrdeepNain', 'displayname': 'Pra...   
2  {'username': 'parmarmaninder', 'displayname': ...   
3  {'username': 'anmoldhaliwal', 'displayname': '...   
4  {'username': 'KotiaPreet', 'displayname': 'Pre...   

                                            outlinks  \
0  [https://twitter.com/ravisinghka/status/136415...   
1                                                 []   
2                                                 []   
3                     [https://youtu.be/-bUKumwq-J8]   
4                                                 []   

                 tcooutlinks  replyCount  retweetCount  ...  quoteCount  \
0  [https://t.co/es3kn0IQAF]           0             0  ...           0   
1                         []           0             0  ...           0   
2                         []           0             0  ...           0   
3  [https://t.co/wBPNdJdB0n]           0             0  ...           0   
4                         []           0             0  ...           0   

        conversationId  lang  \
0  1364506249291784198    en   
1  1364506237451313155    en   
2  1364506195453767680    pa   
3  1364350947099484160    en   
4  1364506144002088963   und   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/android" ...   
2  <a href="http://twitter.com/download/android" ...   
3  <a href="https://mobile.twitter.com" rel="nofo...   
4  <a href="http://twitter.com/download/iphone" r...   

                             sourceUrl          sourceLabel  \
0   http://twitter.com/download/iphone   Twitter for iPhone   
1  http://twitter.com/download/android  Twitter for Android   
2  http://twitter.com/download/android  Twitter for Android   
3           https://mobile.twitter.com      Twitter Web App   
4   http://twitter.com/download/iphone   Twitter for iPhone   

                                               media retweetedTweet  \
0                                               None            NaN   
1  [{'thumbnailUrl': 'https://pbs.twimg.com/ext_t...            NaN   
2                                               None            NaN   
3  [{'thumbnailUrl': 'https://pbs.twimg.com/ext_t...            NaN   
4  [{'previewUrl': 'https://pbs.twimg.com/media/E...            NaN   

                                         quotedTweet  \
0  {'url': 'https://twitter.com/RaviSinghKA/statu...   
1                                               None   
2                                               None   
3                                       

In [5]:
df.describe()

id     replyCount   retweetCount      likeCount  \
count  1.174070e+05  117407.000000  117407.000000  117407.000000   
mean   1.362062e+18       0.631410       5.045619      14.609674   
std    1.293803e+15       9.436159      58.237370     229.083983   
min    1.360040e+18       0.000000       0.000000       0.000000   
25%    1.360935e+18       0.000000       0.000000       0.000000   
50%    1.361945e+18       0.000000       0.000000       1.000000   
75%    1.363097e+18       0.000000       1.000000       2.000000   
max    1.364506e+18    1291.000000    7723.000000   27888.000000   

          quoteCount  conversationId  retweetedTweet  
count  117407.000000    1.174070e+05             0.0  
mean        0.336241    1.361764e+18             NaN  
std         4.474576    1.294352e+16             NaN  
min         0.000000    1.981606e+10             NaN  
25%         0.000000    1.360887e+18             NaN  
50%         0.000000    1.361914e+18             NaN  
75%         0.000000    1.363068e+18             NaN  
max       756.000000    1.364506e+18             NaN

## Pre-procesamiento Data

En esta seccion se abora una etapa de exploracion y pre-procesamiento de la data, para entender sus valores, normalizar la data, hacer limpieza de ser necesario y demas, una vez la etapa de exploracion sea concluida, podemos crear un modulo, de ser necesario, encargado de ejecutar las funciones necesarias para el pre-procesamiento de los datos

#### Punto 1.

- Para abarcar el punto 1 se requiere trabajar con la columna 'date' y 'user' para extraer el top 10 de fechas con mas tweets y el ususario con mas publicaciones 

Iniciaremos trabajando con la columna date, procesandola en pandas y convirtiendola en formato DataTime.

**Nota:** Todo el analisis exploratorio se realizaran en pandas, no se hara pensado en optimizar memoria debido a que es el primer acercamiento a la data, analizando las problematicas y columnas involucradas en cada una, para la deteccion de algun tipo de transfomacion en la data de ser requerido, esta exploracion no es la solucion final de cada uno de los puntos, esta solucion sera realizada en los modulos de python correspondientes, sin embargo se deja el analisis dentro del notebook por motivos de seguimiento de pensamiento logico de la solucion. 

In [6]:
df['newDate']= pd.to_datetime(df['date']).dt.date

df['newDate']

0         2021-02-24
1         2021-02-24
2         2021-02-24
3         2021-02-24
4         2021-02-24
             ...    
117402    2021-02-12
117403    2021-02-12
117404    2021-02-12
117405    2021-02-12
117406    2021-02-12
Name: newDate, Length: 117407, dtype: object

*Utilizacion de funcion **value_counts** de pandas para realizar un conteo de la frecuencia de cada una de las fechas y asi optener el top 10*

In [7]:
top_dates= df['newDate'].value_counts().head(10).index 

In [8]:
top_dates

Index([2021-02-12, 2021-02-13, 2021-02-17, 2021-02-16, 2021-02-14, 2021-02-18,
       2021-02-15, 2021-02-20, 2021-02-23, 2021-02-19],
      dtype='object', name='newDate')

In [9]:
df['user'][0]

{'username': 'ArjunSinghPanam',
 'displayname': 'Arjun Singh Panam',
 'id': 45091142,
 'description': 'Global Citizen, Actor, Director: Sky is the roof above my head, the world is the road I travel, love is my food & mother earth is my bed. Roy in @CosmosMovie',
 'rawDescription': 'Global Citizen, Actor, Director: Sky is the roof above my head, the world is the road I travel, love is my food & mother earth is my bed. Roy in @CosmosMovie',
 'descriptionUrls': [],
 'verified': False,
 'created': '2009-06-06T07:50:57+00:00',
 'followersCount': 603,
 'friendsCount': 311,
 'statusesCount': 17534,
 'favouritesCount': 4269,
 'listedCount': 23,
 'mediaCount': 1211,
 'location': '',
 'protected': False,
 'linkUrl': 'https://www.cosmosmovieofficial.com',
 'linkTcourl': 'https://t.co/3uaoV3gCt3',
 'profileImageUrl': 'https://pbs.twimg.com/profile_images/1215541746492461056/3De61YoQ_normal.jpg',
 'profileBannerUrl': 'https://pbs.twimg.com/profile_banners/45091142/1612601766',
 'url': 'https://twit

*Se observa que la columna **user** es un diccionario de datos, donde el indice **username** contiene el nombre del usuario. Con este hallazgo en mente, podemos filtrar el DataFrame por una de las fechas dentro del top 10, obteniendo asi los tweets para ese dia en especifico, luego hacer uso de la funcion **lambda** de python para recorrer cada usuario dentro de la columna **user**, extrayendo su nombre y realizando un conteo de frecuencia, de nuevo con la funcion **value_counts***

In [10]:
sub_df= df[df['newDate'] == top_dates[0]]
top_user2 = sub_df['user'].apply(lambda x: x['username']).value_counts()

In [11]:
top_user2

user
RanbirS00614606    176
ammiedgr8          116
rebelpacifist      109
lovehazran1         94
Davinder_001        82
                  ... 
imayanmajumdar       1
Nirbhanyadaviyc      1
AnoopChanot          1
HindiNews18          1
rickyrickstir        1
Name: count, Length: 5817, dtype: int64

In [12]:
sub_df

url  \
105060  https://twitter.com/Aman96442728/status/136037...   
105061  https://twitter.com/Gurpreetd86/status/1360378...   
105062  https://twitter.com/Gurpreetd86/status/1360378...   
105063  https://twitter.com/Gurpreetd86/status/1360377...   
105064  https://twitter.com/gpd_bhangra/status/1360377...   
...                                                   ...   
117402  https://twitter.com/rickyrickstir/status/13600...   
117403  https://twitter.com/PunjabTak/status/136004014...   
117404  https://twitter.com/ish_kayy/status/1360040134...   
117405  https://twitter.com/TV9Bharatvarsh/status/1360...   
117406  https://twitter.com/SikhVibes/status/136004012...   

                            date  \
105060 2021-02-12 23:59:59+00:00   
105061 2021-02-12 23:59:57+00:00   
105062 2021-02-12 23:59:40+00:00   
105063 2021-02-12 23:59:16+00:00   
105064 2021-02-12 23:59:07+00:00   
...                          ...   
117402 2021-02-12 01:37:02+00:00   
117403 2021-02-12 01:36:53+00:00   
117404 2021-02-12 01:36:50+00:00   
117405 2021-02-12 01:36:49+00:00   
117406 2021-02-12 01:36:49+00:00   

                                                  content  \
105060                                    #FarmersProtest   
105061  #nodeepkaur #freejagginow #MahapanchayatRevolu...   
105062  @BaazNewsOrg #nodeepkaur #freejagginow #Mahapa...   
105063  #nodeepkaur #freejagginow #MahapanchayatRevolu...   
105064            #FarmersProtest https://t.co/5PArZdboy0   
...                                                   ...   
117402  #FarmersProtest #KisanAndolan #KisaanMajdoorEk...   
117403  PM मोदी की अपील के बीच संयुक्त किसान मोर्चा का...   
117404  United we stand.\nDivided we fall\n#Mahapancha...   
117405  सिंघु बॉर्डर पर लंबी लड़ाई की तैयारी, किसानों ...   
117406  @Kisanektamorcha We are with you, keep the mor...   

                                          renderedContent  \
105060                                    #FarmersProtest   
105061  #nodeepkaur #freejagginow #MahapanchayatRevolu...   
105062  @BaazNewsOrg #nodeepkaur #freejagginow #Mahapa...   
105063  #nodeepkaur #freejagginow #MahapanchayatRevolu...   
105064            #FarmersProtest https://t.co/5PArZdboy0   
...                                                   ...   
117402  #FarmersProtest #KisanAndolan #KisaanMajdoorEk...   
117403  PM मोदी की अपील के बीच संयुक्त किसान मोर्चा का...   
117404  United we stand.\nDivided we fall\n#Mahapancha...   
117405  सिंघु बॉर्डर पर लंबी लड़ाई की तैयारी, किसानों ...   
117406  @Kisanektamorcha We are with you, keep the mor...   

                         id  \
105060  1360378149003874305   
105061  1360378137046118400   
105062  1360378065499602947   
105063  1360377968699338754   
105064  1360377930040430594   
...                     ...   
117402  1360040182771163138   
117403  1360040146402373637   
117404  1360040134230556678   
117405  1360040127679000577   
117406  1360040127146430470   

                                                     user  \
105060  {'username': 'Aman96442728', 'displayname': 'A...   
105061  {'username': 'Gurpreetd86', 'displayname': 'Gu...   
105062  {'username': 'Gurpreetd86', 'displayname': 'Gu...   
105063  {'username': 'Gurpreetd86', 'displayname': 'Gu...   
105064  {'username': 'gpd_bhangra', 'displayname': 'Ga...   
...                                                   ...   
117402  {'username': 'rickyrickstir', 'displayname': '...   
117403  {'username': 'PunjabTak', 'displayname': 'Punj...   
117404  {'username': 'ish_kayy', 'displayname': 'ishy'...   
117405  {'username': 'TV9Bharatvarsh', 'displayname': ...   
117406  {'username': 'SikhVibes', 'displayname': 'Sikh...   

                                                 outlinks  \
105060                                                 []   
105061  [https://twitter.com/UK51NGH/status/1360377290...   
105062                                                 []   
105063  [https://twitter.com/BaazNewsOrg/status/136030...   
105064     

*Ahora podemos realizar un pequeña generalizacion de la solucion, para obtener el top 10 de fechas con mas tweets y los usuario con mas tweets en ese dia, para ello podemos hacer uso del ciclo for:*

In [13]:
result =[]
for date in top_dates:
    #filtro el DataFrame con las fechas especificas 
    print(date)
    sub_df= df[df['newDate'] == date]
    print(len(sub_df))
    
    #encontrar el ususario con mas tweets para la fecha dada
    top_user = sub_df['user'].apply(lambda x: x['username']).value_counts().idxmax()
    
    result.append((date, top_user))

result

2021-02-12
12347
2021-02-13
11296
2021-02-17
11087
2021-02-16
10443
2021-02-14
10249
2021-02-18
9625
2021-02-15
9197
2021-02-20
8502
2021-02-23
8417
2021-02-19
8204


[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'rebelpacifist'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 15), 'jot__b'),
 (datetime.date(2021, 2, 20), 'MangalJ23056160'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 19), 'Preetm91')]

#### Punto 2

- Para abarcar el punto 2, se requiere analizar la columna 'content' la cual contiene el texto de tweet, en esta columna podemos encontrar los emojis utilizados en cada uno de los tweets

In [14]:
tweet=df['content'][1]
tweet

"#FarmersProtest \n#ModiIgnoringFarmersDeaths \n#ModiDontSellFarmers \n@Kisanektamorcha \nFarmers constantly distroying crops throughout India. \nReally, it's hearts breaking...we care about our crops like our children. And govt. agriculture minister is laughing on us🚜🌾WE WILL WIN💪 https://t.co/kLspngG9xE"

Para solucionar la problematica, analizando el contenido del texto, se me ocurren dos posibles solciones:

1. Utilizar expresiones regulares para encontrar los emojis, esto implica crear un diccionario con cada una de las expresiones regulares que me definen un emoji.

2. Usar la biblioteca de python llamada 'emoji', es una biblioteca que me ayuda a la deteccion de emojis en un texto, tengo conocimiento de esta biblioteca debido a que en una de mis experiencias laborales trabaje en un proyecto de scraping y analisis de comentarios en Instagram, esta biblioteca en su momento abarcaba una mayor cantidad de emojis de las que se pueden generalizar por regex, obteniendo un muy buen performance. 

In [15]:
#expresiones regulares 
import re

# Expresión regular para encontrar emojis
emoji_pattern = re.compile("["
                           u"\U0001F600-\U0001F64F"  # emoticons
                           u"\U0001F300-\U0001F5FF"  # symbols & pictographs
                           u"\U0001F680-\U0001F6FF"  # transport & map symbols
                           u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                           "]+", flags=re.UNICODE)

df['emojis'] = df['content'].apply(lambda x: re.findall(emoji_pattern, x))

In [16]:
df['emojis'][1]

['🚜🌾', '💪']

In [17]:
all_emojis = df['emojis'].explode()
all_emojis

0         NaN
1          🚜🌾
1           💪
2         NaN
3         NaN
         ... 
117402    NaN
117403    NaN
117404    NaN
117405    NaN
117406      💪
Name: emojis, Length: 122083, dtype: object

In [18]:
all_emojis[pd.notna(all_emojis)]

1         🚜🌾
1          💪
21        🙄🙄
29        🇮🇳
41        👇👇
          ..
117344     🌾
117368     🛑
117371     💯
117395    🇮🇳
117406     💪
Name: emojis, Length: 18539, dtype: object

In [19]:
all_emojis.value_counts().head(10)

emojis
🙏     2123
😂      633
🌾      605
💚      533
👍      471
👉      450
🇮🇳     426
🙏🙏     403
👇      390
🏽      332
Name: count, dtype: int64

In [57]:
!pip install emoji

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 431.4/431.4 kB 8.7 MB/s eta 0:00:00a 0:00:01


In [20]:
import emoji

# Función para extraer emojis
def extract_emojis(text):
    return [char for char in text if emoji.is_emoji(char)]

In [21]:
# Extraer y contar emojis
df['emojis'] = df['content'].apply(extract_emojis)
all_emojis = df['emojis'].explode()

In [22]:
all_emojis

0         NaN
1           🚜
1           🌾
1           💪
2           🤫
         ... 
117402    NaN
117403    NaN
117404    NaN
117405    NaN
117406      💪
Name: emojis, Length: 146482, dtype: object

In [23]:
all_emojis[pd.notna(all_emojis)]

1         🚜
1         🌾
1         💪
2         🤫
2         🤫
         ..
117371    💯
117374    ✅
117374    ✅
117398    ✊
117406    💪
Name: emojis, Length: 45636, dtype: object

In [24]:
top_emojis = all_emojis.value_counts().head(10)
top_emojis

emojis
🙏    7286
😂    3072
🚜    2972
✊    2411
🌾    2363
🏻    2080
❤    1779
🤣    1668
🏽    1218
👇    1108
Name: count, dtype: int64

#### Punto 3

Para el punto 3 debemos obtener el top de usuarios mas influyendes, dependiendo del conteo de menciones que realiza el usuario en un tweet. Para ello se utiliza la columna **content** para obtener el contenido del tweet y contar el numero de menciones por tweets que realiza el usuario, adcional necesitaremos la columna **user** para obtener el nombre del usuario que realiza la mencion.

Exploraremos la data y una posible solucion, en busca de anomalias en la data y conocimiento de la misma, este analisis continua siendo parte del analisis exploratorio

In [64]:
df['content'][7]

'#ModiDontSellFarmers\n#FarmersProtest https://t.co/uGQb1O5Jg9'

In [33]:
# Expresión regular para encontrar menciones
mention_pattern = re.compile(r'@(\w+)')
df['mentions'] = df['content'].apply(lambda x: mention_pattern.findall(x))

In [46]:
# Crear un contador para menciones hechas por cada usuario
df['username'] = df['user'].apply(lambda x: x['username'])
user_mention_counts = df.groupby('username')['mentions'].sum()

In [52]:
user_count=df.explode('mentions').groupby('username')['mentions'].count()

In [65]:
top_users = user_count.nlargest(10)
list(top_users.items())
top_users

username
loyal90901246      2479
_gurviir           1826
neetuanjle_nitu    1796
JaiHind58236291    1696
preetysaini321      930
JoyJ69957841        915
Gurpreetd86         844
jot__b              840
ScitaramSays        807
mani262002          778
Name: mentions, dtype: int64

In [62]:
list(top_users.items())

[('loyal90901246', 2479),
 ('_gurviir', 1826),
 ('neetuanjle_nitu', 1796),
 ('JaiHind58236291', 1696),
 ('preetysaini321', 930),
 ('JoyJ69957841', 915),
 ('Gurpreetd86', 844),
 ('jot__b', 840),
 ('ScitaramSays', 807),
 ('mani262002', 778)]

## SOLUCION PROBLEMATICA

In [1]:
from q1_time import *


In [2]:
file_path= 'data/farmers-protest-tweets-2021-2-4.json'

## Punto 1 solucion:

Para la solucion del punto 1 partimos del analisis generado en la seccion anterior, para ello lo primero que vamos a realizar es una funcion en pandas, teniendo en cuentas lo alineamientos anteriores, la cual me va obtener las 10 fechas con mas tweets, junto con el usuario que mas tweets realizo para esa fecha.

Para la construccion de esta funcion se utilizo pandas debido a que es un libreria construida para trabajar con volumenes de datos, la idea de utilizar pandas es hacer uso de sus operaciones vectorizadas que se ejecutan rapidamente sobre el Dataframe, asi como la opcion de filtrado eficiente que ofrece esta libreria.

La metodologia a seguir es la siguiente:

1. Se construira una funcion generica para resolver la problematica en base al estudio realizado con anterioridad. Esta funcion a pesar de estar escrita con buenas practicas de uso de la libreria pandas, no esta optimizada y nos servira como base, para optimizar la funcion. 

2. Para la optimizacion de tiempo, se desglozara la funcion generica y se medira los tiempos de ejecucion en cada una de sus partes.

3. Se haran mejoras de optimizacion de tiempo en la funcion, dando como resultado la funcion mejorada, comparada con la funcion original.

4. Para la optimizacion de memoria al igual que para el anterior caso, se analizara y medira el uso de memoria en la funcion y se encontraran puntos a mejorar dentro de ella.

5. Se aplicaran las mejoras de optimizacion de memoria en la funcion, dando como resultado la funcion optimizada. 

#### Funcion Base

La funcion base se realizara y analizara dentro de este notebook, por facilidad de analisis, la funcion final, sera la que disponibilizaremos en el modulo **q1_time.py** y **q1_memory.py** respectivamente

In [3]:
from typing import List, Tuple
from datetime import datetime
import pandas as pd
from joblib import Parallel, delayed
import time

In [7]:
def q1_generic(file_path: str) -> List[Tuple[datetime.date, str]]:
    
    start_time = time.time()
    df = pd.read_json(file_path, lines = True)
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Tiempo de ejecución de lectura: {execution_time} segundos")
    
    start_time2 = time.time()
    df['newDate']= pd.to_datetime(df['date']).dt.date
    
    top_dates= df['newDate'].value_counts().head(10).index 
    result =[]
    for date in top_dates:
        #filtro el DataFrame con las fechas especificas
        sub_df= df[df['newDate'] == date]

        #encontrar el ususario con mas tweets para la fecha dada
        top_user = sub_df['user'].apply(lambda x: x['username']).value_counts().idxmax()

        result.append((date, top_user))
    end_time2 = time.time()
    execution_time2 = end_time2 - start_time2
    print(f"Tiempo de ejecución de procesamiento: {execution_time2} segundos")
    return result

#### Optimizacion de tiempo de ejecucion

In [5]:
result= q1_generic(file_path)

Tiempo de ejecución de lectura: 23.400452375411987 segundos
Tiempo de ejecución de procesamiento: 0.3873133659362793 segundos


In [7]:
%timeit result2= q1_generic(file_path)

Tiempo de ejecución de lectura: 10.13318157196045 segundos
Tiempo de ejecución de procesamiento: 0.3672335147857666 segundos
Tiempo de ejecución de lectura: 8.634846210479736 segundos
Tiempo de ejecución de procesamiento: 0.3708813190460205 segundos
Tiempo de ejecución de lectura: 7.701025724411011 segundos
Tiempo de ejecución de procesamiento: 0.5652155876159668 segundos
Tiempo de ejecución de lectura: 7.95357084274292 segundos
Tiempo de ejecución de procesamiento: 0.3720278739929199 segundos
Tiempo de ejecución de lectura: 7.6641247272491455 segundos
Tiempo de ejecución de procesamiento: 0.37113070487976074 segundos
Tiempo de ejecución de lectura: 7.648245573043823 segundos
Tiempo de ejecución de procesamiento: 0.36950087547302246 segundos
Tiempo de ejecución de lectura: 7.595869541168213 segundos
Tiempo de ejecución de procesamiento: 0.37344861030578613 segundos
Tiempo de ejecución de lectura: 7.617130517959595 segundos
Tiempo de ejecución de procesamiento: 0.36878442764282227 segun

De la prueba anterior tenemos una conclusion clara, y es que el mayor tiempo de procesamiento lo esta tomando la lectura del archivo. Esto hace sentido debido a que para el procesamiento estamos haciendo uso de operaciones vectorizadas de pandas, las cuales son muy optimas para trabajar con grandes volumenes de datos. Por otro lado para la lectura de archivo estamos usando la funcion **read_json** de pandas, la cual no es la manera mas optmiza de realizar la lectura.

Para optimizar el tiempo de ejecucion realizaremos una optmizacion para la lectura del archivo, comparando varias tecnicas como el particionamiento y lectura por chunks, asi como tambien haremos uso de la libreria multiprocessing, para reaizar paralelismo en busca de utilizar al maximo los recursos de CPU.

Lo primero es intentar leer el archivo con otras librerias como dask o vaex

In [8]:
import dask.dataframe as dd
import vaex

/opt/conda/lib/python3.10/site-packages/dask/dataframe/__init__.py:42: FutureWarning: 
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)


In [9]:
%timeit df = pd.read_json(file_path, lines = True)

7.91 s ± 303 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [10]:
%timeit df2 = dd.read_json(file_path).compute()

21.3 s ± 647 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [13]:
%timeit df3 = vaex.from_json(file_path, lines = True)

8.97 s ± 610 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Como se observa el mejor performance para la lectura del json sigue siendo la libreria pandas, por consiguiente haremos uso de otras tecnicas.

haremos la lectura en Chunks, para ello haremos uso de la libreria **orjson** que es significativamente mas rapida que la libreria **json** estandar, para cargar datos de tipo JSON esto reducidara el tiempo de lectura.

Por otro lado hacer la lectura por chunks evita cargar el archivo en memoria de una vez, permitiendo el procesamiento mas rapido y eficiente. Procesar en chunks equilibria bien la carga de I/O y la memoria utilizada.

In [4]:
import orjson

def load_json_in_chunks(file_path: str, chunk_size=10000):
    """
    Carga un archivo JSON en chunks para optimizar el tiempo y uso de memoria.
    
    :param file_path: Ruta del archivo JSON.
    :param chunk_size: Tamaño del chunk en número de líneas.
    :return: Generador que produce chunks del archivo JSON.
    """
    with open(file_path, 'r') as f:
        chunk = []
        for line in f:
            chunk.append(orjson.loads(line))
            if len(chunk) == chunk_size:
                yield chunk
                chunk = []
        if chunk:
            yield chunk

In [16]:
import orjson

start_time = time.time()

chunks = []
for chunk in load_json_in_chunks(file_path):
    df_chunk = pd.DataFrame(chunk)
    chunks.append(df_chunk)
df = pd.concat(chunks, ignore_index=True)

end_time = time.time()
execution_time = end_time - start_time
print(f"Tiempo de ejecución de lectura: {execution_time} segundos")

Tiempo de ejecución de lectura: 6.024820327758789 segundos


In [18]:
%%timeit

chunks = []
for chunk in load_json_in_chunks(file_path):
    df_chunk = pd.DataFrame(chunk)
    chunks.append(df_chunk)
df2 = pd.concat(chunks, ignore_index=True)

3.77 s ± 318 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Como podemos observar, utilizando chunks y la libreria orjson, mejoramos significativamente el tiempo de lectura y ejecucion, en mas de un 50%.

Ahora que pasa si hacemos uso del paralelismo, ejecutando en paralelo cada chunk??

In [19]:
def process_chunk(chunk):
    """
    Procesa un chunk de datos convirtiéndolo en un DataFrame y ajustando la columna de fecha.
    
    :param chunk: Lista de registros JSON.
    :return: DataFrame procesado.
    """
    df_chunk = pd.DataFrame(chunk)
    df_chunk['date'] = pd.to_datetime(df_chunk['date']).dt.date
    return df_chunk

In [20]:
from multiprocessing import Pool, cpu_count

start_time = time.time()

chunks = list(load_json_in_chunks(file_path))
with Pool(cpu_count()) as pool:
    df_chunks = pool.map(process_chunk, chunks)

# Concatenar los DataFrames procesados
df = pd.concat(df_chunks, ignore_index=True)

end_time = time.time()
execution_time = end_time - start_time
print(f"Tiempo de ejecución de lectura: {execution_time} segundos")

Tiempo de ejecución de lectura: 23.689802169799805 segundos


**Interpretacion**: a pesar de que el metodo de paralelismo por teoria, deberia darnos un mejor performance, la sobrecarga de la creacion de procesos y la comunicacion entre ellos, probablemente este causando un rendimiento peor que el esperado. Cabe recalcar que el uso de paralelismo es comendable cuando estamos trabajando con cantidades de datos muy grandes, ya hablando de big data, para este caso, a pesar de que el volumen de datos es alto, este no supera el orden de los MB.


**Conclusion:** En conclusion el uso de multiprocessing en este caso no está dando los beneficios esperados y está introduciendo una sobrecarga significativa. La carga en chunks sin paralelismo es la más eficiente en términos de tiempo de ejecución total.

#### Final Test

In [2]:
from q1_time import q1_time

In [3]:
%timeit result3= q1_time(file_path)

5.17 s ± 314 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [5]:
result3 = q1_time(file_path)
result3

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'rebelpacifist'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 15), 'jot__b'),
 (datetime.date(2021, 2, 20), 'MangalJ23056160'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 19), 'Preetm91')]

### Optimizarion Memoria


Para la optimizacion de memoria, realizaremos un analisis similar al anterior, partiendo de la funcion generic y encontrando puntos a mejorar

In [8]:
def q1_generic(file_path: str) -> List[Tuple[datetime.date, str]]:
    
    df = pd.read_json(file_path, lines = True)
    
    df['newDate']= pd.to_datetime(df['date']).dt.date
    
    top_dates= df['newDate'].value_counts().head(10).index 
    result =[]
    for date in top_dates:
        #filtro el DataFrame con las fechas especificas
        sub_df= df[df['newDate'] == date]

        #encontrar el ususario con mas tweets para la fecha dada
        top_user = sub_df['user'].apply(lambda x: x['username']).value_counts().idxmax()

        result.append((date, top_user))
    return result

In [9]:
%load_ext memory_profiler

%memit q1_generic(file_path)

peak memory: 3188.89 MiB, increment: 2186.48 MiB


#### Logica.

Lo primero que realizaremos para la optimizacion de Memoria es en un inicio mantener la lectura del archivo por chunks, ya que este metodo optimiza en gran medida el uso de memoria al cargar el archivo por lotes, es una de las tecnicas mas conocidad para la optimizacion de Memoria.

Segundo, Evaluaremos el tipo de datos del dataframe, del analisis exploratorio observamos que la mayoria de datos es object, un tipo de dato no muy eficiente si queremos optimizar memoria, para este caso podriamos utilizar un tipo de dato **category** por ejemplo.


Tercero, evitaremos el uso de columnas que no se necesitan dentro de la solucion, asi como la creacion de columnas temporales, que tampoco son necesarias.

In [17]:
def process_chunk(chunk):
    """
    Procesa un chunk de datos convirtiéndolo en un DataFrame y ajustando la columna de fecha.
    Convierte las columnas a tipos de datos más eficientes.
    
    :param chunk: Lista de registros JSON.
    :return: DataFrame procesado.
    """
    df_chunk = pd.DataFrame(chunk)
    
    # Convertir la columna 'date' a datetime y luego a solo fecha
    df_chunk['date'] = pd.to_datetime(df_chunk['date']).dt.date
    
    # Convertir 'user' a un DataFrame separado
    user_df = pd.json_normalize(df_chunk['user'])
    
    # Solo mantener las columnas necesarias
    df_chunk = df_chunk[['date']]
    # Optimizar los tipos de datos
    df_chunk['user'] = user_df['username'].astype('category')
    
    
    return df_chunk



In [18]:
chunks = []
for chunk in load_json_in_chunks(file_path):
    df_chunk = process_chunk(chunk)
    chunks.append(df_chunk)
df = pd.concat(chunks, ignore_index=True)


In [25]:
df.head()

date             user
0  2021-02-24  ArjunSinghPanam
1  2021-02-24       PrdeepNain
2  2021-02-24   parmarmaninder
3  2021-02-24    anmoldhaliwal
4  2021-02-24       KotiaPreet

Con esta optimizacion ya podemos crear una funcion temporal para analizar su optimizacion de memoria:

In [26]:
def q1_memory_temporal(file_path: str) -> List[Tuple[datetime.date, str]]:
    
    chunks = []
    for chunk in load_json_in_chunks(file_path):
        df_chunk = process_chunk(chunk)
        chunks.append(df_chunk)
    df = pd.concat(chunks, ignore_index=True)
    df['user'] = df['user'].astype('category')
    
    # Obtener las 10 fechas con más tweets
    top_dates = df['date'].value_counts().head(10).index
    
    result = []
    for date in top_dates:
        sub_df = df[df['date'] == date]
        
        top_user = sub_df['user'].value_counts().idxmax()
        
        result.append((date, top_user))
    
    return result

In [27]:
%memit q1_memory_temporal(file_path)

peak memory: 1232.96 MiB, increment: 0.00 MiB


#### Conclusion:

Podemos observar una mejora significativa del uso de memoria, de aproximadamente el 30%  pasando de 3188 MB a 1200 MB, Mostrando la viabilidad de este desarrollo. La solucion final la podremos encontrar en el modulo **q1_memory.py**

#### Final test

In [28]:
from q1_memory import q1_memory

In [29]:
%memit q1_memory(file_path)

peak memory: 1232.99 MiB, increment: 0.00 MiB


## Punto 2 Solucion

Para la solucion del punto 2 utilizaremos una logica similar a la que usamos en el punto 1, vamos a crear una funcion generica a partir del analisis realizado en la exploracion de la data, esta funcion generica tendra la logica que resolvera el desafio propuesto, pero esta lejos de ser optimizada tanto en tiempos como en memoria.

Para ello mediremos el tiempo de ejecucio y el uso de memoria en la funcion generacia, y buscaremos optimizar dicha funcion.

In [36]:
import emoji

# Función para extraer emojis
def extract_emojis(text):
    return [char for char in text if emoji.is_emoji(char)]


def q2_generic(file_path: str) -> List[Tuple[str, int]]:
    
    start_time = time.time()
    df = pd.read_json(file_path, lines = True)
    
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Tiempo de ejecución de lectura: {execution_time} segundos")
    
    start_time2 = time.time()
    # Extraer y contar emojis
    df['emojis'] = df['content'].apply(extract_emojis)
    all_emojis = df['emojis'].explode()
    top_emojis = all_emojis.value_counts().head(10)
    
    end_time2 = time.time()
    execution_time2 = end_time2 - start_time2
    print(f"Tiempo de ejecución de proceso: {execution_time2} segundos")
    return list(top_emojis.items())

In [39]:
%%time 
emojis = q2_generic(file_path)

Tiempo de ejecución de lectura: 11.100608825683594 segundos
Tiempo de ejecución de proceso: 4.589721918106079 segundos
CPU times: user 14.9 s, sys: 1.47 s, total: 16.4 s
Wall time: 16.4 s


In [35]:
emojis

[('🙏', 7286),
 ('😂', 3072),
 ('🚜', 2972),
 ('✊', 2411),
 ('🌾', 2363),
 ('🏻', 2080),
 ('❤', 1779),
 ('🤣', 1668),
 ('🏽', 1218),
 ('👇', 1108)]

#### Optimizacion de tiempo de ejecucion

De la ejecucion de la funcion generica, podemos deducir los siguiente puntos:


1. Al igual como sucedio con el punto 1, el mayor tiempo de ejecucion lo esta tomando la lectura del archivo, el cual se esta leyendo haciendo uso de la biblioteca python. Para la optimiazacion de la lectura, haremos uso del analisis previamente realizado, en el cual se concluyo que la manera mas optima de leer el archivo es usando chunks.

2. Podemos observar un tiempo de ejecucion considerable en el proceso, esto es debido a que a pesar de que estamos haciendo buen uso de funciones vectorizadas de pandas, tambien estamos haciendo uso de la libreria emojis, la cual por cada tweet me recorre los caracteres en busca de un emoji, haciendo de este un proceso lento y tedioso si tenemos en cuenta la cantidad de data que tenemos.

**posible solucion apartado 2:** Sabemos que la extraccion de emojis con **emoji.is_emoji** en cada carecter es inificiente, esta problematica se podria solucionar utilizando expresiones regualares que detecten emojis en el textos, haciendolo mas eficiente debido a que no recorre caracter por caracter. Pero si somos fieles al analisis realizado con anterioridad, nos damos cuenta que el uso de expresiones regulares no es una solucion viable en cuanto a calidad del dato y resultado, debido a que tiene problemas para identificar algunos emojis especificos, emojis consecutivos y demas.

En base al analisis realizado en la etapa de exploracion, nos vemos en la necesidad de optimizar el codigo con la biblioteca emoji, debido a que tiene un mejor performance o accuracy. Para ello La función **extract_emojis** se aplica a cada chunk de datos de manera vectorizada utilizando pandas. Esta metodología es eficiente porque el uso de orjson para leer el archivo JSON en chunks reduce significativamente el tiempo de lectura y manejo de memoria, mientras que pandas permite aplicar la función de extracción de emojis de manera vectorizada, mejorando la velocidad de procesamiento sin comprometer la precisión. Esta combinación asegura un rendimiento óptimo al manejar grandes volúmenes de datos.

In [42]:
def process_chunk(chunk):
    """
    Procesa un chunk de datos convirtiéndolo en un DataFrame y extrayendo emojis.
    
    :param chunk: Lista de registros JSON.
    :return: DataFrame procesado con una columna de emojis.
    """
    df_chunk = pd.DataFrame(chunk)
    
    # Extraer emojis del contenido
    df_chunk['emojis'] = df_chunk['content'].apply(extract_emojis)
    
    return df_chunk

In [43]:
def q2_time_temporal(file_path):
  
    start_time = time.time()
    
    # Procesar los chunks y concatenar
    chunks = []
    for chunk in load_json_in_chunks(file_path):
        df_chunk = process_chunk(chunk)
        chunks.append(df_chunk)
    df = pd.concat(chunks, ignore_index=True)
    
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Tiempo de ejecución de lectura y procesamiento inicial: {execution_time} segundos")
    
    start_time2 = time.time()
    # Explode y contar emojis
    all_emojis = df['emojis'].explode()
    top_emojis = all_emojis.value_counts().head(10)
    
    end_time2 = time.time()
    execution_time2 = end_time2 - start_time2
    print(f"Tiempo de ejecución de conteo: {execution_time2} segundos")
    
    return list(top_emojis.items())

In [45]:
%%time 
emojis2= q2_time_temporal(file_path)

Tiempo de ejecución de lectura y procesamiento inicial: 9.614159345626831 segundos
Tiempo de ejecución de conteo: 0.03318309783935547 segundos
CPU times: user 9.92 s, sys: 237 ms, total: 10.2 s
Wall time: 10.2 s


#### conlclusion:

La optimización implementada redujo significativamente el tiempo de lectura del archivo JSON y el tiempo total de ejecución. La lectura en chunks con orjson y el procesamiento vectorizado con pandas mejoraron el rendimiento, reduciendo el tiempo total de ejecución de 16.4 segundos a 10.2 segundos. La combinación de estas técnicas no solo hizo el proceso más rápido sino también más eficiente en términos de manejo de memoria, manteniendo la precisión en la detección de emojis.

Dado lo anterior podemos implementar una solucion general en el modulo **q2_time.py**

### Final Test

In [6]:
from q2_time import q2_time

In [9]:
%%time
final_emoji = q2_time(file_path)

CPU times: user 9.71 s, sys: 448 ms, total: 10.2 s
Wall time: 10.2 s


In [8]:
%timeit q2_time(file_path)

8 s ± 375 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### Optimizacion Memoria q2

Similar al anterior caso, partiremos de la optimizacion de la funcion generic

In [10]:
import emoji

# Función para extraer emojis
def extract_emojis(text):
    return [char for char in text if emoji.is_emoji(char)]


def q2_generic(file_path: str) -> List[Tuple[str, int]]:
    
    df = pd.read_json(file_path, lines = True)
   
    # Extraer y contar emojis
    df['emojis'] = df['content'].apply(extract_emojis)
    all_emojis = df['emojis'].explode()
    top_emojis = all_emojis.value_counts().head(10)

    return list(top_emojis.items())

In [11]:
%load_ext memory_profiler

%memit q2_generic(file_path)

peak memory: 3199.75 MiB, increment: 2397.07 MiB


Para la optimizacion de memoria, utilizaremos el siguiente proceso:
    
1. Lectura selectiva de columnas por chunk: Vamos a leer solo la columna content utilizando orjson y procesaremos solo los datos necesarios.

2. Procesamiento en Chunks: Mantener la lectura en chunks para no cargar todo el archivo en memoria de una vez.

3. Descarte de Datos Innecesarios: Descartar columnas innecesarias inmediatamente después de su uso reduciendo el uso de memoria. En este caso eliminaremos la columna 'content'

4. Uso de pandas y Explosión de Emojis: Utilizamos las capacidades de pandas para operaciones vectorizadas es eficiente y manejable.

In [12]:
def process_chunk(chunk):
    """
    Procesa un chunk de datos extrayendo solo la columna 'content' y los emojis.
    
    :param chunk: Lista de registros JSON.
    :return: DataFrame procesado con una columna de emojis.
    """
    contents = [record['content'] for record in chunk]
    df_chunk = pd.DataFrame(contents, columns=['content'])
    
    # Extraer emojis del contenido
    df_chunk['emojis'] = df_chunk['content'].apply(extract_emojis)
    
    # Descartar la columna 'content' para reducir el uso de memoria
    df_chunk.drop(columns=['content'], inplace=True)
    
    return df_chunk

def q2_memory_temporal(file_path: str) -> List[Tuple[str, int]]:
  
    # Procesar los chunks y concatenar solo la columna de emojis
    chunks = []
    for chunk in load_json_in_chunks(file_path):
        df_chunk = process_chunk(chunk)
        chunks.append(df_chunk)
    df = pd.concat(chunks, ignore_index=True)
    
    # Explode y contar emojis
    all_emojis = df['emojis'].explode()
    top_emojis = all_emojis.value_counts().head(10)
    
    return list(top_emojis.items())

In [13]:
%memit q2_memory_temporal(file_path)

peak memory: 1066.84 MiB, increment: 1.14 MiB


#### Conclusion:

Podemos observar una mejora significativa del uso de memoria, de aproximadamente el 33%  pasando de 3200 MB a 1066 MB, Mostrando la viabilidad de este desarrollo. La solucion final la podremos encontrar en el modulo **q2_memory.py**

### Final Test

In [14]:
from q2_memory import q2_memory

In [15]:
%memit q2_memory(file_path)

peak memory: 1065.98 MiB, increment: 0.00 MiB


## Punto 3 Solucion

In [24]:
import re
def q3_generic(file_path: str) -> List[Tuple[str, int]]:
    
    start_time = time.time()
    df = pd.read_json(file_path, lines = True)
    
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Tiempo de ejecución de lectura: {execution_time} segundos")
    
    start_time2 = time.time()
    # Expresión regular para encontrar menciones
    mention_pattern = re.compile(r'@(\w+)')
    df['mentions'] = df['content'].apply(lambda x: mention_pattern.findall(x))
    
    # Crear un contador para menciones hechas por cada usuario
    df['username'] = df['user'].apply(lambda x: x['username'])
    user_count=df.explode('mentions').groupby('username')['mentions'].count()
    top_users = user_count.nlargest(10)

    end_time2 = time.time()
    execution_time2 = end_time2 - start_time2
    print(f"Tiempo de ejecución de proceso: {execution_time2} segundos")
    
    return list(top_users.items())

In [25]:
%%time
mentions = q3_generic(file_path)

Tiempo de ejecución de lectura: 13.685860872268677 segundos
Tiempo de ejecución de proceso: 1.5820708274841309 segundos
CPU times: user 14.5 s, sys: 1.62 s, total: 16.1 s
Wall time: 16.1 s


In [27]:
# Medicion uso de memoria
%memit q3_generic(file_path)

Tiempo de ejecución de lectura: 19.05904531478882 segundos
Tiempo de ejecución de proceso: 1.495091438293457 segundos
peak memory: 4480.96 MiB, increment: 1166.48 MiB


#### Analisis:

Del punto 1 y 2 hemos aprendido varias cosas a cerca de la data y como optimizarla, asi como tambien diferentes tecnicas de optimizacion tanto de memoria como de tiempo. Para este punto vamos a utilizar todas estas tecnicas que me resultaron exitosas en los puntos anteriores, para abordar la solucion de manera mas eficiente.

Primero que todo el tema de la lectura del archivo: En todos los 3 casos hemos visto que el mayor tiempo de ejecucion lo toma la lectura del archivo, y desde el analisis exhaustivo realizado en el punto 1, nos dimos cuenta que la mejor solucion para optimizar este tiempo, es el uso de lectura por chunks. Es decir, para este punto continuaremos con la misma solucion.

Segundo, nos hemos dado cuenta que no toda la data es necesaria para resolver las problematicas, y que podemos leer el json por chunks, seleccionando las columnas especificas que se necesitan para cada solucion. En este caso, para el punto 3, las columnas *content* & *user*.

Tercero, hemos notado que el uso de funciones vectorizadas es la mejor opcion de optimizacion de tiempo y memoria que podemos tener cuando trabajamos con datos en pandas. En este caso tenemos multiples usos de la funcion **apply** lo cual no es muy optimo. Para este caso, buscare optimizar el uso de este apply en la lectura del json por chunks, al momento de procesarlos.


La solucion para la optimizacion de tiempo es la siguiente:

In [28]:
import re

# Expresión regular para encontrar menciones
mention_pattern = re.compile(r'@(\w+)')

def extract_mentions(text):
    return mention_pattern.findall(text)

def process_chunk(chunk):
    """
    Procesa un chunk de datos extrayendo solo las columnas 'content' y 'user' y las menciones.
    
    :param chunk: Lista de registros JSON.
    :return: DataFrame procesado con una columna de menciones.
    """
    contents = [(record['content'], record['user']['username']) for record in chunk]
    df_chunk = pd.DataFrame(contents, columns=['content', 'username'])
    
    # Extraer menciones del contenido
    df_chunk['mentions'] = df_chunk['content'].apply(extract_mentions)
    
    return df_chunk

Creamos la funcion process_chunk, con la cual aprovechamos el uso de lectura por chunk para extraer solo la columna 'content' & 'user' y adicional, extraemos y creamos la nueva columna mentions, la cual contiene una lista de las menciones realizadas por un usuario para ese tweet

In [29]:
def q3_time_temporal(file_path: str) -> List[Tuple[str, int]]:

    start_time = time.time()
    
    # Procesar los chunks y concatenar
    chunks = []
    for chunk in load_json_in_chunks(file_path):
        df_chunk = process_chunk(chunk)
        chunks.append(df_chunk)
    df = pd.concat(chunks, ignore_index=True)
    
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Tiempo de ejecución de lectura y procesamiento inicial: {execution_time} segundos")
    
    start_time2 = time.time()
    # Crear un contador para menciones hechas por cada usuario
    user_count = df.explode('mentions').groupby('username')['mentions'].count()
    top_users = user_count.nlargest(10)
    
    end_time2 = time.time()
    execution_time2 = end_time2 - start_time2
    print(f"Tiempo de ejecución de conteo: {execution_time2} segundos")
    
    return list(top_users.items())

In [32]:
%%time
mentions2 = q3_time_temporal(file_path)

Tiempo de ejecución de lectura y procesamiento inicial: 6.7898993492126465 segundos
Tiempo de ejecución de conteo: 0.2587771415710449 segundos
CPU times: user 6.98 s, sys: 155 ms, total: 7.13 s
Wall time: 7.13 s


### Conclusion

Podemos ver una gran mejora en el tiempo de ejecucion utilizando la optimizacion propuesta por chunk, y el uso adecuado de las funciones vectorizadas de pandas. La reduccion de tiempo fue de mas del 50%, pasando de un wall time de 17 s a un wall time de 7,13 s

### Final test

In [1]:
from q3_time import q3_time

In [5]:
%%time
mentions3 = q3_time(file_path)

CPU times: user 6.91 s, sys: 475 ms, total: 7.39 s
Wall time: 7.39 s


### Optimizacion de Memoria

Para la optimizacion de memoria al igual que en los casos anteriores, haremos uso de la lectura por chunks como lo hemos trabajado, Adicional usaremos solo las columnas y datos necesrarios para resolver la problematica, reduciendo significativamente el uso de memoria al no usar datos innecesarios y finalmente convertiremos los datos de tipo **object** a **category** el cual es una buena practica para disminuir significativamente el uso de memoria.

In [10]:
import re

# Expresión regular para encontrar menciones
mention_pattern = re.compile(r'@(\w+)')

def extract_mentions(text):
    return mention_pattern.findall(text)

def process_chunk(chunk):
    """
    Procesa un chunk de datos extrayendo solo las columnas 'content' y 'user' y las menciones.
    
    :param chunk: Lista de registros JSON.
    :return: DataFrame procesado con una columna de menciones.
    """
    contents = [(record['content'], record['user']['username']) for record in chunk]
    df_chunk = pd.DataFrame(contents, columns=['content', 'username'])
    
    # Extraer menciones del contenido
    df_chunk['mentions'] = df_chunk['content'].apply(extract_mentions)
    
    # Convertir la columna 'username' a tipo 'category' para ahorrar memoria
    df_chunk['username'] = df_chunk['username'].astype('category')
    
    # Descartar la columna 'content' para reducir el uso de memoria
    df_chunk.drop(columns=['content'], inplace=True)
    
    return df_chunk

En el procesamiento de chunks estamos creando 2 nuevas columnas, la primera es la columna **mentions** la cual contiene las menciones realizadas en un tweet, y segundo la columna **username** la cual es creada a partir de la columna **user** para obtener el usuario que realizo las menciones.

Finalmente eliminamos la columna **content** debido a que no va ser usada en los procesos posteriores, optimizando asi el uso de memoria, mantiendo solo la data 100% necesaria

In [11]:
def q3_memory_temporal(file_path: str) -> List[Tuple[str, int]]:

    # Procesar los chunks y concatenar solo las columnas necesarias
    chunks = []
    for chunk in load_json_in_chunks(file_path):
        df_chunk = process_chunk(chunk)
        chunks.append(df_chunk)
    df = pd.concat(chunks, ignore_index=True)
    
    # Explode y contar menciones
    all_mentions = df['mentions'].explode()
    user_count = all_mentions.groupby(df['username']).count()
    top_users = user_count.nlargest(10)
    
    return list(top_users.items())

In [12]:
%load_ext memory_profiler
# Medicion uso de memoria
%memit q3_memory_temporal(file_path)

The memory_profiler extension is already loaded. To reload it, use:
  %reload_ext memory_profiler
peak memory: 531.88 MiB, increment: 145.05 MiB


### Conclusion:

Podemos observar una optimizacion significativa del uso de memoria, pasando de 4480 MB a tan solo 531 MB, concluyendo que las tecnicas utilizadas dentro de todo el proceso son optimas y aplican muy bien al caso de estudio

### Final test

In [14]:
from q3_memory import q3_memory

In [15]:
%memit q3_memory(file_path)

peak memory: 531.12 MiB, increment: 73.78 MiB
